### DISCUSSION TITLE
Can a download rate from long-term s3 object storage be managed when compacting data?

### DISCUSSION THREAD

**Discussion by bobykus31** [thanos-io/thanos#7221] [Questions & Answers]
We use sefl-hosted s3 object storage for long-stored metrics.  Recently we found the storage overloaded when the download rate exceeds 200MB/sec. A higher download rate leads to maxing out CPU usage.  

![Screenshot from 2024-03-21 11-43-42](https://github.com/thanos-io/thanos/assets/38345247/6f910a4a-b57f-4e4c-87d3-c5da1f969e97)

Unfortunately reducing parallelism did not help much. Or at least the options 

`--wait --wait-interval=1m
--block-viewer.global.sync-block-timeout=30m 
--consistency-delay=2h --log.level=debug 
--block-files-concurrency=48 --block-meta-fetch-concurrency=48 
--compact.blocks-fetch-concurrency=2 
--compact.concurrency=8 
--downsample.concurrency=8 
 --hash-func=SHA256 --compact.enable-vertical-compaction`

Is there a proper way to limit thanos compactors download rate with its settings and still have data compacted/downsampled?

**Comment by yeya24** [comment]
Have you tried lowering `--compact.concurrency=8` and  ` --downsample.concurrency=8`?

**Comment by bobykus31** [comment]
Oh, I thought this settings for processing data already downloaded to the --data-dir, thank you for the tip, I'll try it definitely.

**Comment by bobykus31** [comment]
So, running with` --compact.concurrency=4` and  `--downsample.concurrency=4` does not make any difference.   When concurrency and other related settings are 1, the rate is acceptable (up to 200Mb/s). Now the question is if it can catch up with incoming data. But of course, this is beyond this topic. Thank you for your help, @yeya24! You are the best, as always ;)

**Reply by yeya24** [reply]
Other than this one, I see you already tuned `block-files-concurrency` and `compact.blocks-fetch-concurrency`. Lowering these two should help as well.
